components/core/src/clp/StreamingReader.cpp

```diff
@@ -0,0 +1,518 @@
+
+#include <curl/curl.h>
+
+#include "ErrorCode.hpp"
+#include "Thread.hpp"
+#include "TraceableException.hpp"
+
+namespace clp {
+namespace {
+/**
+ * The exception thrown by a failed libcurl operation.
+ */
+class CurlOperationFailed : public TraceableException {
+public:
+    CurlOperationFailed(
+            ErrorCode error_code,
+            char const* const filename,
+            int line_number,
+            CURLcode err
+    )
+            : m_curl_err{err},
+              TraceableException(error_code, filename, line_number) {}
+
+    [[nodiscard]] auto get_curl_err() const -> CURLcode { return m_curl_err; }
+
+    [[nodiscard]] auto what() const noexcept -> char const* override {
+        return "clp::<unnamed>::CurlDownloadHandler operation failed.";
+    }
+
+private:
+    CURLcode m_curl_err;
+};
+
+/**
+ * This class wraps the C implementation of the Curl string list. It provides a cleaner interface to
+ * manage the life cycle of the object with proper error handling.
+ */
+class CurlStringList {
+public:
+    // Constructor
+    CurlStringList() = default;
+
+    // Copy/Move constructors
+    CurlStringList(CurlStringList const&) = delete;
+    CurlStringList(CurlStringList&&) = delete;
+    auto operator=(CurlStringList const&) -> CurlStringList& = delete;
+    auto operator=(CurlStringList&&) -> CurlStringList& = delete;
+
+    // Destructor
+    ~CurlStringList() {
+        if (nullptr != m_list) {
+            curl_slist_free_all(m_list);
+        }
+    }
+
+    // Methods
+    /**
+     * Appends a string to the end of the list.
+     * @param str
+     * @throw CurlOperationFailed if append operation failed.
+     */
+    auto append(std::string_view str) -> void {
+        auto* list_after_appending{curl_slist_append(m_list, str.data())};
+        if (nullptr == list_after_appending) {
+            throw CurlOperationFailed(ErrorCode_Failure, __FILE__, __LINE__, CURLE_OUT_OF_MEMORY);
+        }
+        m_list = list_after_appending;
+        ++m_size;
+    }
+
+    [[nodiscard]] auto get_raw_list() const -> struct curl_slist* { return m_list; }
+
+    [[nodiscard]] auto get_list_size() const -> size_t { return m_size; }
+
+    [[nodiscard]] auto is_empty() const -> bool { return 0 == get_list_size(); }
+
+private:
+    size_t m_size{0};
+    struct curl_slist* m_list{nullptr};
+};
+
+/**
+ * This class wraps the C implementation of the Curl handler to perform data downloading. It
+ * provides a cleaner interface to manage the life cycle of the object with proper error handling.
+ */
+class CurlDownloadHandler {
+public:
+    using ProgressCallbackType
+            = std::function<int(void*, curl_off_t, curl_off_t, curl_off_t, curl_off_t)>;
+    using WriteCallbackType = std::function<size_t(char*, size_t, size_t, void*)>;
+
+    // Constructor
+    CurlDownloadHandler(
+            StreamingReader& reader,
+            std::string_view src_url,
+            long connection_timeout,
+            long operation_timeout,
```

Comment:
https://curl.se/libcurl/c/CURLOPT_TIMEOUT.html
It is a curl parameter that has a default value. Assuming your data set is huge, and the consuming sleep is slow. You may want to set a large value in case the transfer is killed by CURL.